test/xpu_api/complex/complex_fnc_real.pass.cpp

```diff
@@ -0,0 +1,165 @@
+//      
+template <typename TErrorEngine, typename IsSupportedDouble, typename IsSupportedLongDouble>
+class TestComplexFncReal
+{
+public:
+
+    TestComplexFncReal(TErrorEngine& ee)
+        : errorEngine(ee)
+    {
+    }
+
+    template <typename RunOnHost>
+    void run_test(RunOnHost /*runOnHost*/)
+    {
+        test_fnc_real<float>();
+        TestUtils::invoke_test_if<IsSupportedDouble>()([&]() { test_fnc_real<double>(); });
+        TestUtils::invoke_test_if<IsSupportedLongDouble>()([&]() { test_fnc_real<long double>(); });
+    }
+
+protected:
+
+    template <typename T>
+    void test_fnc_real()
+    {
+        test_fnc_real_form1<T>();
+        test_fnc_real_form2<T>();
+        test_fnc_real_form2<int>();     // DoubleOrInteger, result type checked
+    }
+
+    template <typename T>
+    void test_fnc_real_form1()
+    {
+        test_fnc_real_form1_until_CPP14<T>();
+        test_fnc_real_form1_since_CPP14<T>();
+    }
+
+    template <typename T>
+    void test_fnc_real_form2()
+    {
+        test_fnc_real_form2_until_CPP14<T>();
+        test_fnc_real_form2_since_CPP14<T>();
+    }
+
+    template <typename T>
+    void test_fnc_real_form1_until_CPP14()
+    {
+#if __cplusplus < 201402L
+        const dpl::complex<T> cv(TestUtils::Complex::InitConst<T>::kPartReal);
+        EXPECT_TRUE_EE(errorEngine, TestUtils::Complex::InitConst<T>::kPartReal == dpl::real(cv), "Wrong effect of dpl::real() #1");
+#endif
+    }
+
+    template <typename T>
+    void test_fnc_real_form1_since_CPP14()
+    {
+#if __cplusplus >= 201402L
+        COMPLEX_TEST_CONSTEXPR dpl::complex<T> cv(TestUtils::Complex::InitConst<T>::kPartReal);
+        EXPECT_TRUE_EE(errorEngine, TestUtils::Complex::InitConst<T>::kPartReal == dpl::real(cv), "Wrong effect of dpl::real() #2");
+#endif
+    }
+
+    template <typename T>
+    void test_fnc_real_form2_until_CPP14()
+    {
+#if __cplusplus < 201402L
+        const T z = TestUtils::Complex::InitConst<T>::kPartReal;
+        const auto real_res = dpl::real(z);
+        EXPECT_EQ_TYPE_EE(errorEngine, typename TestUtils::Complex::InitConst<T>::DestComplexFieldType, real_res);
+        
+        EXPECT_TRUE_EE(errorEngine, TestUtils::Complex::InitConst<T>::kExpectedResReal == real_res, "Wrong effect of dpl::real() #3");
+#endif
+    }
+
+    template <typename T>
+    void test_fnc_real_form2_since_CPP14()
+    {
+#if __cplusplus >= 201402L
+        COMPLEX_TEST_CONSTEXPR T z = TestUtils::Complex::InitConst<T>::kPartReal;
+        COMPLEX_TEST_CONSTEXPR auto real_res = dpl::real(z);
+        EXPECT_EQ_TYPE_EE(errorEngine, typename TestUtils::Complex::InitConst<T>::DestComplexFieldType, real_res);
+
+        EXPECT_TRUE_EE(errorEngine, TestUtils::Complex::InitConst<T>::kExpectedResReal == real_res, "Wrong effect of dpl::real() #4");
+#endif
+    }
+
+private:
+
+    TErrorEngine& errorEngine;
+};
+
+int
+main()
+{
+    bool bSuccess = true;
+
+    if (!TestUtils::Complex::test_on_host<TestComplexFncReal>())
```

Comment:
Better `bool bSuccess = TestUtils::Complex::test_on_host<TestComplexFncReal>();` 

instead of 133-136 lines.

Reply:
Fixed